python/GafferUI/PlugPopup.py

```diff
@@ -95,9 +85,113 @@ def __keyPress( self, widget, event ) :
+		script = sole( p.ancestor( Gaffer.ScriptNode ) for p in plugs )
+		assert( script is not None )
+
+		column = GafferUI.ListContainer( spacing = 4 )
+		_PopupWindow.__init__( self, "", child = column, **kw )
+
+		with column :
+
+			# Title
+
+			if title is None :
+
+				# Make default title
+
+				if len( plugs ) == 1 :
+					title = plugs[0].relativeName( script )
+				else :
+					nodes = { plug.node() for plug in plugs }
+					title = "{} plugs{}".format(
+						len( plugs ),
+						" on {} nodes".format( len( nodes ) ) if len( nodes ) > 1 else ""
+					)
+
+					commonNode = plugs[0].node()
+					for plug in plugs :
+						if not commonNode.isAncestorOf( plug ) :
+							commonNode = commonNode.commonAncestor( plug.node() )
+
+					if commonNode != script :
+						title = "{} ({})".format(
+							commonNode.relativeName( script ),
+							title
+						)
+
+			if title :
+
+				titleWidget = GafferUI.Label( "<h4>{}</h4>".format( title ) )
+				titleWidget.setToolTip(
+					"\n".join( "- " + p.relativeName( script ) for p in plugs )
+				)
+
+			# Widget row
+
+			with GafferUI.ListContainer( spacing = 4, orientation = GafferUI.ListContainer.Orientation.Horizontal ) :
+
+				# Warning
+
+				if warning :
+					warningBadge = GafferUI.Image( "warningSmall.png" )
+					warningBadge.setToolTip( warning )
+
+				# PlugValueWidget, or label explaining why we can't show one
+
+				if sole( p.__class__ for p in plugs ) :
+					self.__plugValueWidget = GafferUI.PlugValueWidget.create( plugs )
+				else :
+					self.__plugValueWidget = None
+					GafferUI.Label( "Unable to edit plugs with mixed types" )
+
+		self.visibilityChangedSignal().connect( Gaffer.WeakMethod( self.__visibilityChanged ), scoped = False )
+
+	def popup( self, center = None ) :
+
+		_PopupWindow.popup( self, center )
+
+		# Attempt to focus the first text widget. This is done after making
+		# the window visible, as we check child widget visibility to avoid
+		# attempting to focus hidden widgets.
+
+		textWidget = self.__textWidget( self.__plugValueWidget )
+		if textWidget is not None :
+			if isinstance( textWidget, GafferUI.TextWidget ) :
+				textWidget.grabFocus()
+				textWidget.setSelection( 0, len( textWidget.getText() ) )
+			else :
+				textWidget.setFocussed( True )
+			textWidget._qtWidget().activateWindow()
+
+	def plugValueWidget( self ) :
+
+		return self.__plugValueWidget
+
+	def __visibilityChanged( self, unused ) :
+
+		if self.visible() :
+			return
+
+		# The modal popup will hide the window without de-focusing the
+		# current widget. Ultimately this results in un-committed edits
+		# from being lost. Force the active widget to resign focus.
+		focusWidget = QtWidgets.QApplication.focusWidget()
+		if focusWidget is not None :
+			gafferWidget = GafferUI.Widget._owner( focusWidget )
+			if gafferWidget is not None and self.isAncestorOf( gafferWidget ) :
+				gafferWidget._qtWidget().clearFocus()
+
 	@classmethod
 	def __textWidget( cls, plugValueWidget ) :
```

Comment:
Not significant, but I would find this slightly clearer as __firstTextWidget if this is just used in the one place in popup()

Reply:
89e842a